### PR TITLE
New Resource: aws_ecrpublic_repository_policy

### LINKED ISSUE #16540: Support Managing ECR Public Repository Policy
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

ECR Public was announced at re:Invent 2020. This API is separate from the existing ECR (Private) API. Support managing a repository and its associated catalog data.

### New or Affected Resource(s)

* aws_ecrpublic_repository

### Potential Terraform Configuration

```hcl
resource "aws_ecrpublic_repository_policy" "example" {
  registry_id = ""
  repository_name = aws_ecrpublic_repository.example.repository_name
  policy_text = ""
}
```

### References

* [Announcement](https://aws.amazon.com/about-aws/whats-new/2020/12/announcing-amazon-ecr-public-and-amazon-ecr-public-gallery/)
* [Product Page](https://gallery.ecr.aws/)
* [API Reference](https://docs.aws.amazon.com/AmazonECRPublic/latest/APIReference/Welcome.html)
* [AWS Go SDK Service Reference](https://docs.aws.amazon.com/sdk-for-go/api/service/ecrpublic/)

### PR DESCRIPTION
**Pull request by breathingdust**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #16540

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* **New Resource:** `aws_ecrpublic_repository_policy` 
```

Output from acceptance testing:

```console
make testacc TEST=./aws TESTARGS='-run=TestAccAWSEcrPublicRepositoryPolicy_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEcrPublicRepositoryPolicy_ -timeout 180m
=== RUN   TestAccAWSEcrPublicRepositoryPolicy_basic
=== PAUSE TestAccAWSEcrPublicRepositoryPolicy_basic
=== RUN   TestAccAWSEcrPublicRepositoryPolicy_iam
=== PAUSE TestAccAWSEcrPublicRepositoryPolicy_iam
=== CONT  TestAccAWSEcrPublicRepositoryPolicy_basic
=== CONT  TestAccAWSEcrPublicRepositoryPolicy_iam
--- PASS: TestAccAWSEcrPublicRepositoryPolicy_basic (34.82s)
--- PASS: TestAccAWSEcrPublicRepositoryPolicy_iam (43.79s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       46.488s
```

### CODE REVIEW

**Comment by pgacek** [comment]
Hi - any ideas when it's gonna be released?

**Comment by ewbankkit** [comment]
The `XAT001: missing ErrorCheck` awsproviderlint errors can be addressed by replacing

```go
		PreCheck:     func() { testAccPreCheck(t); testAccPreCheckAwsEcrPublic(t) },
```

with

```go
		PreCheck:     func() { testAccPreCheck(t); testAccPreCheckAwsEcrPublic(t) },
		ErrorCheck:   testAccErrorCheck(t, ecrpublic.EndpointsID),
```

**Comment by ewbankkit** [comment]
Need to add a `.changelog/16901.txt` file.

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+		Create: resourceAwsEcrPublicRepositoryPolicyCreate,
+		Read:   resourceAwsEcrPublicRepositoryPolicyRead,
+		Update: resourceAwsEcrPublicRepositoryPolicyUpdate,
+		Delete: resourceAwsEcrPublicRepositoryPolicyDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"repository": {
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:26]: I would change this to `repository_name` to match the AWS API.

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+		return err
+	}
+
+	if out == nil {
+		return fmt.Errorf("error reading ECR Public Repository Policy: empty response")
+	}
+
+	log.Printf("[DEBUG] Received repository policy %s", out)
+
+	repositoryPolicy := out
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:115]: No need to create another variable here, just use `out` in the `d.Set()` calls.

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+	return nil
+}
+
+func resourceAwsEcrPublicRepositoryPolicyUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ecrpublicconn
+
+	if !d.HasChange("policy") {
+		return nil
+	}
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:130]: This isn't necessary as the Plugin SDK will only call the `Update` handler if `policy` has changed (as long as the other attributes are defined correctly, which they are).

File: website/docs/r/ecrpublic_repository_policy.html.markdown [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+Provides an Elastic Container Registry Public Repository Policy.
+
+Note that currently only one policy may be applied to a repository.
+
+## Example Usage
+
+```hcl
+resource "aws_ecrpublic_repository" "foo" {
+  repository_name = "bar"
+}
```
Comment (ewbankkit) [inline website/docs/r/ecrpublic_repository_policy.html.markdown:20]:
```suggestion
resource "aws_ecrpublic_repository" "example" {
  repository_name = "Example"
}
```

File: website/docs/r/ecrpublic_repository_policy.html.markdown [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+resource "aws_ecr_repository_policy" "foopolicy" {
+  repository = aws_ecrpublic_repository.foo.repository_name
```
Comment (ewbankkit) [inline website/docs/r/ecrpublic_repository_policy.html.markdown:23]:
```suggestion
resource "aws_ecr_repository_policy" "example" {
  repository = aws_ecrpublic_repository.example.repository_name
```

File: website/docs/r/ecrpublic_repository_policy.html.markdown [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+}
+EOF
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `repository` - (Required) Name of the repository to apply the policy.
```
Comment (ewbankkit) [inline website/docs/r/ecrpublic_repository_policy.html.markdown:60]: As suggested above.

```suggestion
* `repository_name` - (Required) Name of the repository to apply the policy.
```

File: website/docs/r/ecrpublic_repository_policy.html.markdown [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+* `policy` - (Required) The policy document. This is a JSON formatted string. For more information about building IAM policy documents with Terraform, see the [AWS IAM Policy Document Guide](https://learn.hashicorp.com/terraform/aws/iam-policy)
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `repository` - The name of the repository.
```
Comment (ewbankkit) [inline website/docs/r/ecrpublic_repository_policy.html.markdown:67]: The repository name is an argument, no need to list as an exported attribute.

```suggestion
```

File: aws/resource_aws_ecrpublic_repository_policy_test.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccAWSEcrPublicRepositoryPolicy_basic(t *testing.T) {
+	randString := acctest.RandString(10)
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy_test.go:16]: I suggest using our de-facto standard of

```go
rName := acctest.RandomWithPrefix("tf-acc-test")
```

for the randomly generated name.

File: aws/resource_aws_ecrpublic_repository_policy_test.go [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+	resourceName := "aws_ecrpublic_repository_policy.default"
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy_test.go:17]: We prefer Terraform resource names to be `test` wherever possible.

```suggestion
	resourceName := "aws_ecrpublic_repository_policy.test"
```

File: aws/resource_aws_ecrpublic_repository_policy_test.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+func testAccAWSEcrPublicRepositoryPolicy(randString string) string {
+	return fmt.Sprintf(`
+resource "aws_ecrpublic_repository" "foo" {
+  repository_name = "tf-acc-test-ecr-%s"
+}
+
+resource "aws_ecrpublic_repository_policy" "default" {
+  repository = aws_ecrpublic_repository.foo.repository_name
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy_test.go:145]: For this and other test configurations:

```suggestion
func testAccAWSEcrPublicRepositoryPolicy(rName string) string {
	return fmt.Sprintf(`
resource "aws_ecrpublic_repository" "test" {
  repository_name = %[1]q
}

resource "aws_ecrpublic_repository_policy" "test" {
  repository_name = aws_ecrpublic_repository.test.repository_name
```

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if isResourceTimeoutError(err) {
+		out, err = conn.SetRepositoryPolicy(&input)
+	}
+	if err != nil {
+		return fmt.Errorf("Error updating ECR Repository Policy: %s", err)
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:158]:
```suggestion
		return fmt.Errorf("error updating ECR Repository Policy (%s): %w", d.Id(), err)
```

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+		RepositoryName: aws.String(d.Id()),
+		RegistryId:     aws.String(d.Get("registry_id").(string)),
+	})
+	if err != nil {
+		if ecrerr, ok := err.(awserr.Error); ok {
+			switch ecrerr.Code() {
+			case "RepositoryNotFoundException", "RepositoryPolicyNotFoundException":
+				return nil
+			default:
+				return err
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:182]:
```suggestion
				return fmt.Errorf("error deleting ECR Public Repository Policy", d.Id(), err)
```

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+	if err != nil {
+		return fmt.Errorf("Error updating ECR Repository Policy: %s", err)
+	}
+
+	repositoryPolicy := *out
+
+	d.SetId(aws.StringValue(repositoryPolicy.RepositoryName))
+	d.Set("registry_id", repositoryPolicy.RegistryId)
+
+	return nil
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:166]: `Update` should call `Read` to get updated values.

```suggestion
	return resourceAwsEcrPublicRepositoryPolicyRead(d, meta)
```

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+		RepositoryName: aws.String(d.Id()),
+	})
+	if err != nil {
+		if ecrerr, ok := err.(awserr.Error); ok {
+			switch ecrerr.Code() {
+			case "RepositoryNotFoundException", "RepositoryPolicyNotFoundException":
+				d.SetId("")
+				return nil
+			default:
+				return err
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:103]:
```suggestion
				return fmt.Errorf("error reading ECR Public Repository Policy (%s): %w", d.Id(), err)
```

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if isResourceTimeoutError(err) {
+		out, err = conn.SetRepositoryPolicy(&input)
+	}
+	if err != nil {
+		return fmt.Errorf("Error creating ECR Public Repository Policy: %s", err)
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:72]:
```suggestion
		return fmt.Errorf("error creating ECR Public Repository Policy: %w", err)
```

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+	if out == nil {
+		return fmt.Errorf("error reading ECR Public Repository Policy: empty response")
+	}
+
+	log.Printf("[DEBUG] Received repository policy %s", out)
+
+	repositoryPolicy := out
+
+	d.SetId(aws.StringValue(repositoryPolicy.RepositoryName))
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:117]: `d.SetId()` should only be called in `Create`

```suggestion
```

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	if err != nil {
+		return fmt.Errorf("Error creating ECR Public Repository Policy: %s", err)
+	}
+
+	if out == nil {
+		return fmt.Errorf("error creating ECR Public Repository Policy: empty response")
+	}
+
+	repositoryPolicy := *out
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:79]: No need to create another variable here, just use `out` in the d.SetId() call.

File: aws/resource_aws_ecrpublic_repository_policy.go [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+
+	log.Printf("[DEBUG] ECR Public repository policy created: %s", *repositoryPolicy.RepositoryName)
+
+	d.SetId(aws.StringValue(repositoryPolicy.RepositoryName))
+	d.Set("registry_id", repositoryPolicy.RegistryId)
```
Comment (ewbankkit) [inline aws/resource_aws_ecrpublic_repository_policy.go:84]: `registry_id` will be set in the `Read` call.

```suggestion
```

**Comment by aquam8** [comment]
Thank you for trying to complete this PR 👍

**Comment by breathingdust** [comment]
Hi all 👋  Just letting you know that this is issue is featured on this quarters roadmap. If a PR exists to close the issue a maintainer will review and either make changes directly, or work with the original author to get the contribution merged. If you have written a PR to resolve the issue please ensure the "Allow edits from maintainers" box is checked. Thanks for your patience and we are looking forward to getting this merged soon!

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% AWS_DEFAULT_REGION=us-east-1 make testacc TESTS=TestAccECRPublicRepositoryPolicy_ PKG=ecrpublic
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecrpublic/... -v -count 1 -parallel 20 -run='TestAccECRPublicRepositoryPolicy_' -timeout 180m
=== RUN   TestAccECRPublicRepositoryPolicy_basic
=== PAUSE TestAccECRPublicRepositoryPolicy_basic
=== RUN   TestAccECRPublicRepositoryPolicy_disappears
=== PAUSE TestAccECRPublicRepositoryPolicy_disappears
=== RUN   TestAccECRPublicRepositoryPolicy_Disappears_repository
=== PAUSE TestAccECRPublicRepositoryPolicy_Disappears_repository
=== RUN   TestAccECRPublicRepositoryPolicy_iam
=== PAUSE TestAccECRPublicRepositoryPolicy_iam
=== CONT  TestAccECRPublicRepositoryPolicy_basic
=== CONT  TestAccECRPublicRepositoryPolicy_iam
=== CONT  TestAccECRPublicRepositoryPolicy_disappears
=== CONT  TestAccECRPublicRepositoryPolicy_Disappears_repository
--- PASS: TestAccECRPublicRepositoryPolicy_Disappears_repository (14.92s)
--- PASS: TestAccECRPublicRepositoryPolicy_disappears (16.57s)
--- PASS: TestAccECRPublicRepositoryPolicy_basic (27.85s)
--- PASS: TestAccECRPublicRepositoryPolicy_iam (40.11s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ecrpublic	46.285s
```

**Pull request merged**